src/ComServerHelpers/Internal/BaseActivationFactoryWrapper.cs

```diff
@@ -0,0 +1,91 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Runtime.InteropServices;
+using System.Runtime.InteropServices.Marshalling;
+using ComServerHelpers.Windows.Com;
+using Windows.Win32.Foundation;
+using static Windows.Win32.PInvoke;
+
+namespace ComServerHelpers.Internal;
+
+[GeneratedComClass]
+internal sealed partial class BaseActivationFactoryWrapper(BaseActivationFactory factory, ComWrappers comWrappers) : IActivationFactory
+{
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("Design", "CA1031:Do not catch general exception types", Justification = "COM method, must not throw.")]
+    public unsafe HRESULT ActivateInstance(void** instance)
+    {
+        if (instance is null)
+        {
+            return HRESULT.E_INVALIDARG;
+        }
+
+        nint unknown = 0;
+        try
+        {
+            object managedInstance = factory.ActivateInstance();
+            unknown = comWrappers.GetOrCreateComInterfaceForObject(managedInstance, CreateComInterfaceFlags.None);
+            *instance = (void*)unknown;
```

Comment:
If the `OnInstanceCreated()` call can throw, this should be placed after that. An `HRESULT` returning function shouldn't update any outs if an error is returned. Imagine the following caller, which is very common in COM infrastructure code.

```cpp
IUnknown* pUnk;
HRESULT hr = ActivateInstance(&pUnk);
if (FAILED(hr))
{
    return hr;
}

// Store the pUnk in a smart pointer here or in the caller.
```

The exceptional case should also `Marshal.Release()` on the non-null `unknown`.

Reply:
Note that fixing the scenario so the `Release` is called makes setting the out an A/V waiting to happen if someone uses a smart COM pointer and will then double release.

Reply:
I don't expect `OnInstanceCreated` to throw ever since it's simply invoking raising an event on `EventHandler`.
But I moved it regardless and add release to unknown if it's non-null.